### PR TITLE
Make hasAll work as expected.

### PR DESCRIPTION
**Pull request by Krijger**
$hasAll now checks if all elements of an array are present.
Before, $hasAll did exactly the same as $has.

I wrote a test for it, and had to fix $all, which was dependent on Hash.hasAll, which had a bug. Did this by adding Hash.hasExactly.

I need some help though. I'm new to node.js and docpad. What I can't figure out just jet is how I can use my altered version of query-engine in a local docpad project. It's more difficult than a normal module, since this is a module for docpad itself.

### CODE REVIEW

File: src/documents/lib/query-engine.js.coffee [hunk 1]
```diff
@@ -1299,7 +1305,7 @@ class Query
 		'$all':
 			test: (opts) ->
 				if opts.selectorValue? and opts.modelValueExists
-					if (new Hash opts.modelValue).hasAll(opts.selectorValue)
```
Comment (balupton) [inline src/documents/lib/query-engine.js.coffee]: Is this how mongodb handles this?

File: src/documents/lib/query-engine.js.coffee [hunk 2]
```diff
@@ -261,9 +261,9 @@ class Hash extends Array
 		pass = true
 
 		# Perform the check
-		for value in @
+		for value in options
 			empty = false
-			unless value in options
```
Comment (balupton) [inline src/documents/lib/query-engine.js.coffee]: Comment should be added to explain this

**Comment by balupton** [comment]
@Krijger @damirarh 

> Is nobody maintaining this project any more? I can't see any other reason why there would be no response to this pull request in almost 2 years. I've stumbled upon the same issue today and had to work around it because of this bug.

At the time of this I went through a mental breakdown. Now there is much more federation with the projects. Sorry for the delay.

**Comment by balupton** [comment]
@blitmap is reviewing now

File: src/documents/lib/query-engine.js.coffee [hunk 3]
```diff
@@ -272,6 +272,12 @@ class Hash extends Array
 		# Return whether we passed or not
 		return pass
 
+	# Has Exactly
+	# Check if all the options exist within us, and no other
+	hasExactly: (options) ->
+		hasAll = @hasAll(options)
+		return hasAll and util.toArray(options).length is @.length
+
```
Comment (balupton) [inline src/documents/lib/query-engine.js.coffee]: This should just be a equality check... which should already be implemented.

**Comment by blitmap** [comment]
After talking about this for a while it seems `$hasAll` is supposed to work identically to `$all`.  And `$all` is currently correct.

We plan to align the existing selectors with Mongo soon - the initial plan is to remove the `$has*` selectors, which (again) should act identically to `$in` and `$all`.

(I appreciate your work here, but we're going to remove the selector you're trying to fix - it's not in Mongo)

**Comment by balupton** [comment]
Closed in favour of https://github.com/bevry/query-engine/issues/44

**Pull request closed without merging**